### PR TITLE
Transformation utility to fix sequence association

### PR DESCRIPTION
**Pull request by rolfhm**
Includes the routine `fix_scalar_syntax` that transforms actual arguments like a(1,1) passed to array dummy arguments into array syntax. It needs some documentation and a test for derived types, so making a draft pull request.

Is the format OK? Should there be some kind of transformation class? Have I put it in the right directory?

### CODE REVIEW

**Comment by rolfhm** [comment]
What's the appropriate way to deal with derived type objects. I treat them as `DeferredTypeSymbol`, but should it be a `TypedSymbol` in case the class definition is available to Loki?

**Comment by mlange05** [comment]
Hi @rolfhm , first of all thanks for looking into this; having a utility mechanism for resolving this type of argument-passing style is certainly very useful and powerful. 

As for the implementation style, I actually think this is already looking pretty good. I personally would prefer all utility routines to have at least small docstrings (I know! :wink:), and I haven't done a thorough code review yet, but for a first implementation this is looking quite ok to me. Yes, eventually we might want to add a `Transformation` class, so that we can run this via the scheduler in its own right, but that doesn't need to to be included in a first version (we have a few pure utility transformations that are implemented purely procedurally.

The one key thing to point out is that you will need some testing for this, as we do not allow individual PRs to degrade test coverage beyond a small threshold. But this can be done with a procedural interface, pretty much as we do for the other utility transformations (try to test it standalone first, not as part of SCC). The final point I noted is that, if you integrate this straight into the SCC transformation, which totally makes sense, it might be good to have a flag to switch this off (a constructor flag to the SCCBaseTransformation, just in case). This can be on by default, but having an off-switch can be useful, just in case things go wrong.

Other than that, this looks like it's well on its way, so please feel free to put this into proper PR review mode once you've added a test or two. :smile:

**Comment by rolfhm** [comment]
Is there any way to run a single test with a cmake install?  `ctest -R` can only narrow it down to `loki_tests`, which is not very useful. If I try to run pytest manually, it can't find loki and when I tried to do a pip install, it failed with an esoteric error message when trying to install numpy.
```
              File "/tmp/pip-build-env-ctsxnf2e/overlay/lib/python3.12/site-packages/pkg_resources/__init__.py", line 2172, in <module>
                register_finder(pkgutil.ImpImporter, find_on_path)
                                ^^^^^^^^^^^^^^^^^^^
            AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
            [end of output]
```

**Comment by rolfhm** [comment]
Spent too much time getting into the details on how to simplify expressions, but I resisted the temptation to try to resolve all the nested Sums. The `simplify_sum` and `product_value` functions might belong in their respective classes with a bit of rework.

If your doing arithmetic operations on the Loki versions of `Sum` and `Product`, you get the Pymbolic versions in return, which can cause unexpected behaviour and might qualify as an issue.

**Review by reuterbal — CHANGES_REQUESTED** [review]
First of all, apologies for the delay in getting to this, and many thanks for this contribution!
I agree with Michael, the implementation looks very good and putting this in as a Loki transformation helper utility does sound look the right place to me. 

My main request would be to not call this "scalar syntax". The correct term in Fortran is "sequence association" and I'd vote that this should be used throughout here (i.e., command-line argument, CMake options and Python code).

The testing is great. Note that you should be able to just compare to strings rather than instances of the relevant expression nodes, which would make the test a bit shorter and more accessible - but that's a minor detail.

The `product_value` utility is currently under-tested (see [test coverage](https://app.codecov.io/gh/ecmwf-ifs/loki/pull/173?src=pr&el=desc#42c0bb88d28d9848edea93c381234a5b-R83)). Would it be possible to add 2-3 more cases to the test which trigger the remaining cases?

**Comment by rolfhm** [comment]
Agree with the name change. I changed the option to `-resolve-sequence-association` and the transformation to `transform_sequence_association`

File: tests/test_transform_sequence_association.py
Comment (rolfhm) [reply tests/test_transform_sequence_association.py:76]: I was looking into it, but discovered that the code crashes if the dimension inside the routine is given by a Sum or a Product, for example `dimension(k+1)` and `dimension(-k)`, respectively.
The solution is to modify the `process_symbol` routine to recursively process such cases, but I'm wondering if I've been  really overthinking the problem. Since sequence association is always contiguous, I think we could just plug in the relevant array dimensions from the caller routine. At least the two calls in this example case produce the same output:
```
program crashdummies

    use blarg, only: blurg

    implicit none

    real, dimension(10,5) :: A
    integer :: j

    A = 0.

    call blurg(A(5:10,1))

    do j = 1,3
        print *, A(:,j)
    enddo
    print *

    A = 0.

    call blurg(A(5:19,1))

    do j = 1,3
        print *, A(:,j)
    enddo

end program crashdummies

module blarg

contains

    subroutine blurg(x)

        implicit none

        real, dimension(15) :: x

        integer :: i

        do i = 1, 15
            x(i) = 1.
        enddo

    end subroutine blurg

end module blarg
```

Would the inliner have any problem with this?

Comment (reuterbal) [reply tests/test_transform_sequence_association.py:46]: Fantastic, simplicity is always better and this looks really neat now! Pylint has flagged a few redundant symbols and imports now.

I don't think the inliner should have a problem with this, at least there is a test that includes ranges: https://github.com/ecmwf-ifs/loki/blob/a2f2428e3f3ee3215f9dee90ae2dc438ba0cb1e3/tests/test_transform_inline.py#L374

**Comment by rolfhm** [comment]
And now it passes the code checks again :)

**Review by mlange05 — APPROVED** [review]
First of all, many thanks for a very comprehensive piece of work! After the clean-up this all looks very well implemented and tested, so all good from my side. Very much appreciate this one, as it's a very useful utility, even outside of the primary target context! :pray: 

Since this the commit history for this one has a few twists and turns, I'd propose doing a squash merge, but other than that, GTG from my side.

**Review by reuterbal — APPROVED** [review]
Excellent, many thanks again!

**Comment by rolfhm** [comment]
Should I do a squash commit manually, or is it possible to do it as part of the merge?

**Comment by reuterbal** [comment]
I'll take care of that, just waiting for the test suite to finish after the last PR merge

**Pull request merged**
